File: courseMaterial/course_ROSActions/code_section_3_simple_actions_activity/my_robot_tutorials/src/count_until_client.cpp
#include <ros/ros.h>
#include <actionlib/client/simple_action_client.h>
#include <actionlib/client/terminal_state.h>

#include <my_robot_msgs/CountUntilAction.h>
#include <my_robot_msgs/CountUntilGoal.h>
#include <my_robot_msgs/CountUntilResult.h>
#include <my_robot_msgs/CountUntilFeedback.h>

class CountUntilClient {

	protected:
	ros::NodeHandle _nh;
	actionlib::SimpleActionClient<my_robot_msgs::CountUntilAction> _ac;

	public:
	CountUntilClient():
		_ac("/count_until", true)
	{
		ROS_INFO("Wait for the action server to start...");
		_ac.waitForServer();
		ROS_INFO("Server is now up.");
	}

	void sendGoal() {
		my_robot_msgs::CountUntilGoal goal;
		goal.max_number = 12;
		goal.wait_duration = 0.3;
		_ac.sendGoal(goal, 
			boost::bind(&CountUntilClient::doneCb, this, _1, _2),
			boost::bind(&CountUntilClient::activeCb, this),
			boost::bind(&CountUntilClient::feedbackCb, this, _1));
		ROS_INFO("Goal has been sent.");

		// Uncomment to cancel the goal after 2 seconds
		//ros::Duration(2.0).sleep();
		//_ac.cancelGoal();
	}

	void doneCb(const actionlib::SimpleClientGoalState& state,
			const my_robot_msgs::CountUntilResultConstPtr &result) 
	{
		ROS_INFO("Finished in state: %s", 
			state.toString().c_str());
		ROS_INFO("Count result: %d", (int)result->count);
	}

	void activeCb() {
		ROS_INFO("Goal just went active");
	}

	void feedbackCb(const my_robot_msgs::CountUntilFeedbackConstPtr &feedback) 
	{
		ROS_INFO("Feedback received. Percentage: %lf", 
			feedback->percentage);
	}
};

int main (int argc, char **argv) {
	ros::init(argc, argv, "count_until_client");
	CountUntilClient client;
	client.sendGoal();
	ros::spin();
}

File: courseMaterial/course_ROSActions/code_section_5_actions_activity/my_robot_tutorials/scripts/count_until_client.py
#!/usr/bin/env python

import rospy
import actionlib

from my_robot_msgs.msg import CountUntilAction
from my_robot_msgs.msg import CountUntilGoal

class CountUntilClient:

    def __init__(self):
        self._ac = actionlib.SimpleActionClient('/count_until',
                CountUntilAction)
        self._ac.wait_for_server()
        rospy.loginfo("Action server is up, we can send new goals!")

    def send_goal_and_get_result(self):
        goal = CountUntilGoal(max_number=18, wait_duration=0.5)
        self._ac.send_goal(goal, done_cb=self.done_callback,
                feedback_cb=self.feedback_callback)
        rospy.loginfo("Goal has been sent.")
       
        #rospy.sleep(2)
        #goal2 = CountUntilGoal(max_number=7, wait_duration=0.5)
        #self._ac.send_goal(goal2, done_cb=self.done_callback,
        #        feedback_cb=self.feedback_callback)
        #rospy.loginfo("Goal has been sent.")

		## Uncomment to cancel the goal after 2 seconds
        #rospy.sleep(2)
        #self._ac.cancel_goal()
        #rospy.loginfo("Send cancel request")

    def done_callback(self, status, result):
        rospy.loginfo("Status is : " + str(status))
        rospy.loginfo("Result is : " + str(result))

    def feedback_callback(self, feedback):
        rospy.loginfo(feedback)

if __name__ == '__main__':
    rospy.init_node('count_until_client')

    client = CountUntilClient()

    client.send_goal_and_get_result()

    rospy.spin()
